src/workerd/api/streams/standard.c++

```diff
@@ -2869,8 +2869,22 @@ public:
         // that the PumpToReader, and the sink it owns, are always accessed from the right
         // IoContext. Thw WeakRef ensures that if the PumpToReader is freed while
         // the JS continuation is pending, there won't be a dangling reference.
-        return ioContext.awaitJs(
-            pumpLoop(js, ioContext, kj::mv(readable), ioContext.addObject(addWeakRef())));
+        //
+        // The call to awaitJs here can throw a JsExceptionThrown error synchronously
+        // if, for instance, the isolate is being terminated due to an exception.
+        // We wrap this in a js.tryCatch to catch that case and return a rejected
+        // kj::Promise instead.
+        return js.tryCatch([&] {
+          return ioContext.awaitJs(
+              pumpLoop(js, ioContext, kj::mv(readable), ioContext.addObject(addWeakRef())));
+        }, [&](jsg::Value exception) {
+          auto handle = exception.getHandle(js);
+          if (handle.IsEmpty()) {
```

Comment:
When can this be empty? By my reading this shouldn't ever happen.

Reply:
Not sure, still investigating that. It absolutely can be.

Reply:
Like I'm pretty sure both that `exception` cannot be empty, but also `getHandle()` cannot possibly return an empty handle -- if `exception` were empty, then `exception.getHandle()` would crash?

Reply:
Ok, more detail here: it seems there are conditions (that I still need to investigate further) under which the `v8::Promise::Then(...)` will return an empty `MaybeLocal` without scheduling an exception on the isolate. In such cases, it always appears that `isolate->IsExecutionTerminating()` is true. The result is that the `check(...)` in `thenImpl` in `promise.h` ends up throwing the `JsExceptionThrown` without having an exception scheduled on the isolate, leading to the empty exception here. We'll need to investigate further the exact issue why `Then` is returning empty without scheduling an exception but the PR handles the initial failure case more gracfully now.